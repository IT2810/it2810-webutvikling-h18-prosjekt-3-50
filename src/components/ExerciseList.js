import React, { Component } from 'react'

import { StyleSheet } from 'react-native'

import {Card, List, ListItem, Text, Left, Body, Right, Button, View, Row, H2, Picker, Icon, Toast } from 'native-base'

export default class ExerciseList extends Component {
  constructor(props, context) {
    super(props, context)
    this.state = {
      exercises: [
        {name: 'Squat', sets: '4', reps: '12'},
        {name: 'Benchpress', sets: '4', reps: '12'},
        {name: 'Pullup', sets: '4', reps: '12'},
        {name: 'Row', sets: '4', reps: '12'}
      ]
    }

    this.showToast = this.showToast.bind(this)
    this._add = this._add.bind(this)
  }
  
  _remove(value) {
    // TODO: delete in redux
    this.setState(prevState => ({
      exercises: prevState.exercises.filter(exercise => exercise.name !== value.name)
    }))

    this.showToast()
  }

  showToast() {
    Toast.show({
      type: 'warning',
      duration: 3000,
      text: 'Deleted exercise'
    })
  }

  _add() {
    this.props.navigation.navigate('AddExercise')
  }

  render () {

    return (
      <View>
        <Row>
          <Left>
            <H2>Exercises</H2>
          </Left>
          <Right>
            <Button
              primary
              block
              onPress={this._add}
              testID={'addExerciseButton'}
            >
              <Text> ADD EXERCISE </Text>
            </Button>
          </Right>
        </Row>

        <List
          dataArray={this.state.exercises}
          renderRow={(exercise) => 
            <ListItem>
              <Left>
                <Text> { exercise.name }  </Text>
              </Left>
              <Body>
                <Text> {exercise.reps} x {exercise.sets} </Text>
              </Body>
              <Right>
                <Button
                  danger
                  onPress={this._remove.bind(this, exercise)}
                >
                  <Icon active name="trash" />
                </Button>
              </Right>
            </ListItem>
          }
        >
        </List>
      </View>
    ) 
  }
}