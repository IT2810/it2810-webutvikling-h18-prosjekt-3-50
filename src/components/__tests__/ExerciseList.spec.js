/**
 * @jest-environment jsdom
 */

import React from 'react'
import { shallow } from 'enzyme'
import ExerciseList from '../ExerciseList'
import toJson from 'enzyme-to-json';
import 'native-base'
import { findByID } from '../../testUtils.js'

describe('ExerciseList', () => {
  let wrapper
  let navigateMock
  
  beforeEach(() => {
    navigateMock = jest.fn()
    const navigation = {navigate: navigateMock}
    wrapper = shallow(<ExerciseList navigation={navigation}/>)
    wrapper.instance().showToast = jest.fn()
  })

  it('renders correctly', () => {
    expect(toJson(wrapper.dive())).toMatchSnapshot()
  })

  describe('_remove', () => {
    it('removes from list', () => {

      wrapper.setState({exercises: [
        {name: 'Squat', sets: '4', reps: '12'},
        {name: 'Benchpress', sets: '4', reps: '12'},
        {name: 'Pullup', sets: '4', reps: '12'},
        {name: 'Row', sets: '4', reps: '12'}
      ]})

      var exerciseRemoved = {name: 'Pullup', sets: '4', reps: '12'}
      wrapper.instance()._remove(exerciseRemoved)

      expect(wrapper.state('exercises').indexOf(exerciseRemoved)).toEqual(-1)
      expect(wrapper.state('exercises')).toHaveLength(3)
    })
  })

  describe('_add', () => {
    it('called when clicking on add exercise button', () => {
      let _addMock = jest.fn()
      wrapper.instance()._add = _addMock
      wrapper.instance().forceUpdate()

      let addExerciseButton = findByID(wrapper, 'addExerciseButton')
      addExerciseButton.props().onPress()

      expect(_addMock.mock.calls.length).toBe(1)
    })
    
    it('navigates to AddExercise', () => {
      wrapper.instance()._add()

      expect(navigateMock.mock.calls.length).toBe(1)
    })

  })
})