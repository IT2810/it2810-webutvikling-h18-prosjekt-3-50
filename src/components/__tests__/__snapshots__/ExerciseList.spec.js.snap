// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExerciseList renders correctly 1`] = `
<ViewNB
  style={Object {}}
>
  <RowNB>
    <Styled(Left)>
      <Styled(Text)
        style={
          Object {
            "fontSize": 30,
            "fontWeight": "600",
            "lineHeight": 45,
          }
        }
      >
        Exercises
      </Styled(Text)>
    </Styled(Left)>
    <Styled(Right)>
      <Styled(Button)
        block={true}
        onPress={[Function]}
        primary={true}
        testID="addExerciseButton"
      >
        <Styled(Text)>
           ADD EXERCISE 
        </Styled(Text)>
      </Styled(Button)>
    </Styled(Right)>
  </RowNB>
  <Styled(Card)
    style={
      Object {
        "marginBottom": "5%",
        "marginTop": "5%",
      }
    }
  >
    <Styled(ListItem)
      key="0"
    >
      <Styled(Left)
        onPress={[Function]}
      >
        <Styled(Text)>
           
          Squat
            
        </Styled(Text)>
      </Styled(Left)>
      <Styled(Body)>
        <Styled(Text)>
           
          12
           x 
          4
           
        </Styled(Text)>
      </Styled(Body)>
      <Styled(Right)>
        <Styled(Button)
          danger={true}
          onPress={[Function]}
        >
          <Styled(Icon)
            active={true}
            name="trash"
          />
        </Styled(Button)>
      </Styled(Right)>
    </Styled(ListItem)>
    <Styled(ListItem)
      key="1"
    >
      <Styled(Left)
        onPress={[Function]}
      >
        <Styled(Text)>
           
          Benchpress
            
        </Styled(Text)>
      </Styled(Left)>
      <Styled(Body)>
        <Styled(Text)>
           
          12
           x 
          4
           
        </Styled(Text)>
      </Styled(Body)>
      <Styled(Right)>
        <Styled(Button)
          danger={true}
          onPress={[Function]}
        >
          <Styled(Icon)
            active={true}
            name="trash"
          />
        </Styled(Button)>
      </Styled(Right)>
    </Styled(ListItem)>
    <Styled(ListItem)
      key="2"
    >
      <Styled(Left)
        onPress={[Function]}
      >
        <Styled(Text)>
           
          Pullup
            
        </Styled(Text)>
      </Styled(Left)>
      <Styled(Body)>
        <Styled(Text)>
           
          12
           x 
          4
           
        </Styled(Text)>
      </Styled(Body)>
      <Styled(Right)>
        <Styled(Button)
          danger={true}
          onPress={[Function]}
        >
          <Styled(Icon)
            active={true}
            name="trash"
          />
        </Styled(Button)>
      </Styled(Right)>
    </Styled(ListItem)>
    <Styled(ListItem)
      key="3"
    >
      <Styled(Left)
        onPress={[Function]}
      >
        <Styled(Text)>
           
          Row
            
        </Styled(Text)>
      </Styled(Left)>
      <Styled(Body)>
        <Styled(Text)>
           
          12
           x 
          4
           
        </Styled(Text)>
      </Styled(Body)>
      <Styled(Right)>
        <Styled(Button)
          danger={true}
          onPress={[Function]}
        >
          <Styled(Icon)
            active={true}
            name="trash"
          />
        </Styled(Button)>
      </Styled(Right)>
    </Styled(ListItem)>
  </Styled(Card)>
</ViewNB>
`;
